### DISCUSSION TITLE
Graceful shutdown?

### DISCUSSION THREAD

**Discussion by StefanNemeth** [prisma/prisma#2895] [Q&A] [answered]
I'd like to shutdown my application gracefully and do some database work before I exit the process myself.

However, Prisma exits the process on a SIGINT event on its own. Even though I got a work-around by monkey-patching "process.exit", the Prisma query engine still exits with the SIGINT event and I'm unable to execute queries therefore. 

`The Node.js process already received a SIGINT signal, therefore the Prisma query engine exited
and your request can't be processed.`

Is there any way to prevent this behaviour?

Thanks in advance.

**Comment by StefanNemeth** [comment, marked as answer]
If that's helpful to anyone, here's the very hacky work-around that I use (though I'd appreciate something else)

```javascript
const realProcessExit = process.exit;
process.exit = () => {
  // Monkey-patch process.exit because Prisma wants to exit before
  // we can shut down gracefully
};

import { PrismaClient } from "@prisma/client";

["EXIT", "SIGINT", "SIGUSR1", "SIGUSR2", "uncaughtException", "SIGTERM"].forEach((eventType) => {
  const previousListeners = process.listeners(eventType);
  process.removeAllListeners(eventType);

  process.once(eventType, async (param) => {
    await shutdown();

    previousListeners.forEach((l) => l(param));

    if (eventType !== "EXIT") {
      realProcessExit();
    }
  });
});
```

**Reply by janpio** [reply]
Can you create an issue for what you would like to see, also include your workaround as "alternatives" and we can make sure to look into this further and possibly prioritize work on it.

**Comment by ludralph** [comment]
Hi @StefanNemeth,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
